### DISCUSSION TITLE
Serial Queue vs Concurrent with barrier tasks

### DISCUSSION THREAD

**Discussion by marcusway** [mxcl/PromiseKit#1278] [Q&A]
I've found `PromiseKit` enormously useful to use, and, as a relative noob in iOS, I've learned a lot by just poking around the source. So thanks to all contributors!

I was looking at the implementation of `when` [here](https://github.com/mxcl/PromiseKit/blob/93c8d41ce96ed78f36c3948be396d76f3ca3de1b/Sources/when.swift#L24), and I was wondering: is using a concurrent `DispatchQueue` and then submitting only `.barrier` tasks to it any different from submitting tasks to a `serial` queue? In other words, would this code behave differently if `barrier` were a serial queue (and you didn't pass the `.barrier` flag to the `sync` call?

```swift
...

let barrier = DispatchQueue(label: "org.promisekit.barrier.when", attributes: .concurrent)

for promise in thenables {
    promise.pipe { result in
        barrier.sync(flags: .barrier) {
             // do stuff
        }
    }
}
```

**Comment by mxcl** [comment]
If we aren’t reading from it (looks like yeah) then probably it is behaving the same as a serial queue, yes. Haven't checked the code though.
